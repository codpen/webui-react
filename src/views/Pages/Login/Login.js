import React, {Component} from 'react';
import {
    Button,
    Card,
    CardBody,
    CardGroup,
    Col,
    Container,
    Form,
    Input,
    InputGroup,
    InputGroupAddon,
    InputGroupText,
    Row,
    UncontrolledAlert
} from 'reactstrap';
import {connect} from "react-redux";
import {changeAuthKey, changeIPAddress, changeUserNamePassword, signOut} from "../../../actions/userActions";
import axiosInstance from "../../../utils/API/API";
import urls from "../../../utils/API/endpoint";


function removeParam(parameter) {
    let url = document.location.href;
    let urlparts = url.split('?');

    if (urlparts.length >= 2) {
        let urlBase = urlparts.shift();
        let queryString = urlparts.join("?");

        let prefix = encodeURIComponent(parameter) + '=';
        let pars = queryString.split(/[&;]/g);
        for (let i = pars.length; i-- > 0;)
            if (pars[i].lastIndexOf(prefix, 0) !== -1)
                pars.splice(i, 1);
        if (pars.length > 0)
            url = urlBase + '?' + pars.join('&');
        else
            url = urlBase;
        window.history.pushState('', document.title, url); // added this line to push the new url directly to url bar .

    }
    return url;
}

class Login extends Component {

    constructor(props) {
        super(props);
        let ipAddress = window.location.href.split("#/")[0];
        if (ipAddress.indexOf("?") !== -1)
            ipAddress = window.location.href.split("?")[0];
        // if (localStorage.getItem(IP_ADDRESS_KEY))
        //     ipAddress = localStorage.getItem(IP_ADDRESS_KEY);
        this.state = {
            username: "",
            password: "",
            ipAddress,
            connectionSuccess: false,
            error: ""
        };
    }

    changeUserName = e => {
        this.setState({
            username: e.target.value,
            connectionSuccess: false
        });
    };
    changePassword = e => {
        this.setState({
            password: e.target.value,
            connectionSuccess: false

        })
    };
    changeIPAddress = e => {

        this.setState({
            ipAddress: e.target.value,
            connectionSuccess: false
        });
    };

    redirectToDashboard = () => {
        this.props.history.push('/dashboard');
    };


    onSubmit = e => {
        if (e)
            e.preventDefault();

        const {ipAddress, username, password} = this.state;

        Promise.all([
            changeUserNamePassword(username, password),
            changeIPAddress(ipAddress)
        ]).then(() => {
            this.redirectToDashboard()
        });

    };

    checkConnection = (e) => {
        e.preventDefault();

        // Set the localStorage parameters temporarily.
        const {ipAddress, username, password} = this.state;
        const {changeUserNamePassword, changeIPAddress} = this.props;

        Promise.all([
            changeUserNamePassword(username, password),
            changeIPAddress(ipAddress)
        ]).then(() => {
            axiosInstance.post(urls.noopAuth).then((data) => {
                console.log("Connection successful.");
                this.setState({
                    connectionSuccess: true,
                    error: ""
                })
            }, (error) => {
                console.log(error);
                this.setState({
                    connectionSuccess: false,
                    error: "Error connecting. Please check username password and verify if rclone is working at the specified IP."
                })
            })
        })


    };

    componentDidMount() {
        localStorage.clear();
        this.props.signOut();


        let url_string = window.location.href;
        let url = new URL(url_string);
        let loginToken = url.searchParams.get("login_token");
        let ipAddress = this.state.ipAddress;
        if (url.searchParams.get("ip_address")) {
            ipAddress = url.searchParams.get("ip_address");
        }
        // console.log(loginToken);
        if (loginToken) {
            Promise.all([
                this.props.changeAuthKey(loginToken),
                this.props.changeIPAddress(ipAddress)
            ]);
            removeParam("login_token");
            removeParam("ip_address");
            this.redirectToDashboard();
        }
    }


    render() {
        const {username, password, ipAddress, connectionSuccess, error} = this.state;

        return (
            <div className="app flex-row align-items-center" data-test="loginComponent">
                <Container>
                    <Row className="justify-content-center">
                        <Col md="8">
                            <CardGroup>
                                <Card className="p-4">
                                    <CardBody>
                                        <Form onSubmit={this.onSubmit}>
                                            <h1>Login</h1>
                                            <p className="text-muted">Sign In to your account</p>
                                            {error && <UncontrolledAlert color="danger" children={error}/>}
                                            {connectionSuccess && <UncontrolledAlert color="success"
                                                                                     children={"Connection verified. You may now login."}/>}
                                            <InputGroup className="mb-3">
                                                <InputGroupAddon addonType="prepend">
                                                    <InputGroupText>
                                                        <i className="icon-user"></i>
                                                    </InputGroupText>
                                                </InputGroupAddon>
                                                <Input type="text" placeholder="IP Address / URL"
                                                       autoComplete="ipAddress"
                                                       onChange={this.changeIPAddress} value={ipAddress}/>
                                            </InputGroup>
                                            <InputGroup className="mb-3">
                                                <InputGroupAddon addonType="prepend">
                                                    <InputGroupText>
                                                        <i className="icon-user"></i>
                                                    </InputGroupText>
                                                </InputGroupAddon>
                                                <Input type="text" placeholder="Username" autoComplete="username"
                                                       onChange={this.changeUserName} value={username}/>
                                            </InputGroup>
                                            <InputGroup className="mb-4">
                                                <InputGroupAddon addonType="prepend">
                                                    <InputGroupText>
                                                        <i className="icon-lock"></i>
                                                    </InputGroupText>
                                                </InputGroupAddon>
                                                <Input type="password" placeholder="Password"
                                                       autoComplete="current-password" onChange={this.changePassword}
                                                       value={password}/>
                                            </InputGroup>
                                            <Row>
                                                <Col xs="6">
                                                    <Button color="primary" className="px-4"
                                                            disabled={!connectionSuccess}>Login</Button>
                                                </Col>
                                                <Col xs="6" className="text-right">
                                                    <Button onClick={this.checkConnection} color="primary"
                                                            className="px-4">Verify</Button>
                                                </Col>
                                            </Row>
                                        </Form>
                                    </CardBody>
                                </Card>
                                <Card className="text-white bg-primary py-5 d-md-down-none" style={{width: '44%'}}>
                                    <CardBody className="text-center">
                                        <div>
                                        </div>
                                    </CardBody>
                                </Card>
                            </CardGroup>
                        </Col>
                    </Row>
                </Container>
            </div>
        );
    }
}


export default connect(null, {signOut, changeUserNamePassword, changeIPAddress, changeAuthKey})(Login);
