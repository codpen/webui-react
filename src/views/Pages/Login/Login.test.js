import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import Login from './Login';
import {shallow} from "enzyme/build";
import {Provider} from "react-redux";
import {testStore} from "../../../../Utils";

const setUp = (props = {}) => {
    const component = shallow(
        <Provider store={testStore()}>
            <MemoryRouter><Login/></MemoryRouter>
        </Provider>
    );
    return component;
}

describe('Login Page', function () {

    let wrapper;
    beforeEach(() => {
        const props = {};
        wrapper = setUp(props)
    });

    it('should render without crashing', function () {
        expect(wrapper).toHaveLength(1)
    });
});