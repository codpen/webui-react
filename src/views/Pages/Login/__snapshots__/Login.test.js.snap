// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remote Explorer Layout renders should match snapshot 1`] = `
<div
  className="app flex-row align-items-center"
  data-test="loginComponent"
>
  <Container
    tag="div"
  >
    <Row
      className="justify-content-center"
      tag="div"
    >
      <Col
        md="8"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <CardGroup
          tag="div"
        >
          <Card
            className="p-4"
            tag="div"
          >
            <CardBody
              tag="div"
            >
              <Form
                onSubmit={[Function]}
                tag="form"
              >
                <h1>
                  Login
                </h1>
                <p
                  className="text-muted"
                >
                  Sign In to your account
                </p>
                <InputGroup
                  className="mb-3"
                  tag="div"
                >
                  <InputGroupAddon
                    addonType="prepend"
                    tag="div"
                  >
                    <InputGroupText
                      tag="span"
                    >
                      <i
                        className="icon-user"
                      />
                    </InputGroupText>
                  </InputGroupAddon>
                  <Input
                    autoComplete="ipAddress"
                    onChange={[Function]}
                    placeholder="IP Address / URL"
                    type="text"
                    value="http://localhost:5572/"
                  />
                </InputGroup>
                <InputGroup
                  className="mb-3"
                  tag="div"
                >
                  <InputGroupAddon
                    addonType="prepend"
                    tag="div"
                  >
                    <InputGroupText
                      tag="span"
                    >
                      <i
                        className="icon-user"
                      />
                    </InputGroupText>
                  </InputGroupAddon>
                  <Input
                    autoComplete="username"
                    onChange={[Function]}
                    placeholder="Username"
                    type="text"
                    value=""
                  />
                </InputGroup>
                <InputGroup
                  className="mb-4"
                  tag="div"
                >
                  <InputGroupAddon
                    addonType="prepend"
                    tag="div"
                  >
                    <InputGroupText
                      tag="span"
                    >
                      <i
                        className="icon-lock"
                      />
                    </InputGroupText>
                  </InputGroupAddon>
                  <Input
                    autoComplete="current-password"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
                  />
                </InputGroup>
                <Row
                  tag="div"
                >
                  <Col
                    tag="div"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                    xs="6"
                  >
                    <Button
                      className="px-4"
                      color="primary"
                      disabled={true}
                      tag="button"
                    >
                      Login
                    </Button>
                  </Col>
                  <Col
                    className="text-right"
                    tag="div"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                    xs="6"
                  >
                    <Button
                      className="px-4"
                      color="primary"
                      onClick={[Function]}
                      tag="button"
                    >
                      Verify
                    </Button>
                  </Col>
                </Row>
              </Form>
            </CardBody>
          </Card>
          <Card
            className="text-white bg-primary py-5 d-md-down-none"
            style={
              Object {
                "width": "44%",
              }
            }
            tag="div"
          >
            <CardBody
              className="text-center"
              tag="div"
            >
              <div />
            </CardBody>
          </Card>
        </CardGroup>
      </Col>
    </Row>
  </Container>
</div>
`;
